### PR TITLE
Optimize DA verify_transactions

### PR DESCRIPTION
**Pull request by yaziciahmet**
# Description

Optimizes verify_transactions method of DA verifier.

Previously:
1. Iterate over completeness proof (Verifying each tx in completeness also exist in inclusion)
  1.1. Validate wtxid
  1.2. Ensure corresponding tx exists in inclusion proof
  1.3. Collect completeness proof wtxids into a BTreeSet
2. Iterate over inclusion proof (Verifying each relevant tx in inclusion also exist in completeness)
  2.1 If prefix matches, verify it also exists in collected completeness proof wtxids BTreeSet and remove it
3. If completeness proof wtxids BTreeSet not empty, there is extra tx in completeness, hence invalid

New:
1. Filter out nonrelevant transactions from inclusion proof
2. ZipEq filtered out inclusion and completeness, they must be of same length
  1.1 Verify wtxids match

With the new version, collecting and later removing wtxids into an BTreeSet is removed, and there is now only 1 iteration over proofs.

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/bitcoin-da/src/verifier.rs [hunk 1]
```diff
@@ -1116,30 +1098,6 @@ mod tests {
         );
     }
 
-    #[test]
```
Comment (eyusufatik) [inline crates/bitcoin-da/src/verifier.rs:1119]: why remove test

Comment (yaziciahmet) [reply crates/bitcoin-da/src/verifier.rs:1119]: if you check closely, you will see that it is the exact same test with: break_completeness_proof_order test

**Comment by yaziciahmet** [comment]
I did some benchmarks locally, and on average optimized version consumes 0.5% less cycles. This test was only with 1 sequencer commitment, hence, I can safely say that more commitments will result in a greater cycle difference.

File: crates/bitcoin-da/src/verifier.rs [hunk 2]
```diff
@@ -82,29 +80,26 @@ impl DaVerifier for BitcoinVerifier {
...
 
-            // make sure completeness txs are ordered same in inclusion proof
-            // this logic always start seaching from the last found index
-            // ordering should be preserved naturally
-            let is_found_in_block =
-                inclusion_iter.any(|wtxid_inc| wtxid_inc == wtxid.as_byte_array());
-            if !is_found_in_block {
-                return Err(ValidationError::RelevantTxNotFoundInBlock);
+            // completeness proof tx for the relevant tx must exist
+            let Some(tx) = completeness_iter.next() else {
```
Comment (kpp) [inline crates/bitcoin-da/src/verifier.rs]: https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.zip_longest

Comment (yaziciahmet) [reply crates/bitcoin-da/src/verifier.rs]: with zip_longest we would have to check if the values are some, imo panicing with zip_eq is better. As there is no recovery we can perform if we gave the wrong input.

File: crates/bitcoin-da/src/verifier.rs (hunk 1 above)
Comment (eyusufatik) [reply crates/bitcoin-da/src/verifier.rs:1119]: let's remove `txs.swap(2, 3);` from break_completeness_proof_order and restore this test then

File: crates/bitcoin-da/src/verifier.rs [hunk 3]
```diff
@@ -82,29 +81,20 @@ impl DaVerifier for BitcoinVerifier {
...
-            let is_found_in_block =
-                inclusion_iter.any(|wtxid_inc| wtxid_inc == wtxid.as_byte_array());
-            if !is_found_in_block {
-                return Err(ValidationError::RelevantTxNotFoundInBlock);
+        let relevant_wtxid_iter = inclusion_proof
+            .wtxids
+            .iter()
+            .filter(|wtxid| wtxid.starts_with(prefix));
+        let completeness_iter = completeness_proof.iter();
+        for (wtxid, tx) in relevant_wtxid_iter.zip_eq(completeness_iter) {
```
Comment (jfldde) [inline crates/bitcoin-da/src/verifier.rs]:
```suggestion
for (wtxid, tx) in relevant_wtxid_iter.zip_eq(&completeness_proof) {
```

**Review by jfldde — APPROVED** [review]

**Pull request merged**